simple metadata format in some transaction endpoints

paolino opened the pull request:
- We implement support for transaction creation (old and new) 
- we implement a query parameter `simple-metadata` on `get transaction` and `list transactions` to select the new  format in the response 
- we implement a switch in the CLI commands get and list to support the new format in the output 
- we use `oneOf ` openapi combinator to document the possibility of the two formats for the same endpoints

### User facing changes

-    implement additional simple metadata format in get and list transaction output (simple-metadata flag)
-    implement additional simple metadata format in transaction creation (no flag needed)


### Issue Number

ADP-1596

jonathanknowles reviewed (changes requested):
Hi @paolino 

Many thanks for making this PR!

I've not had a chance to look through all the functionality yet, but there are a couple of things to fix:

1. The latency benchmark currently doesn't build, because `Link.getTransaction` now takes an extra argument. If you're typically building with `cabal build all`, then you might want to run `cabal configure --enable-benchmarks`. This will make sure that `cabal build all` also builds the benchmarks! (I was also caught out by this the other day...)

2. There are some import and formatting issues that we should fix.

I created PR #3257 with suggested fixes for **_both_** of these issues. If you're okay with the commits in that PR, please feel free to squash them into your own branch, or cherry-pick them if you prefer. 👍🏻

Unisay commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -729,6 +730,12 @@ cmdWalletGetUtxoStatistics mkClient =
     deriving (Show, Eq)
 
+-- | which json schema to use for output, True is simple
+metadataSchemaOption :: Parser Bool
```
May I suggest you using a dedicated sum type for this case (e.g. like the one above) to avoid [boolean blindness](https://existentialtype.wordpress.com/2011/03/15/boolean-blindness/) ?
JFYI there is such package https://hackage.haskell.org/package/choice-0.2.2/docs/Data-Choice.html (I am not suggesting it here)

paolino replied:
Done

Unisay commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1435,16 +1450,16 @@ transactionSubmitPayloadArgument = argumentT $ mempty
 --
 -- Note: we decode the JSON just so that we can validate more client-side.
-metadataOption :: Parser ApiTxMetadata
+metadataOption :: Parser (Maybe TxMetadataWithSchema)
```
Consider changing its type to `Parser TxMetadataWithSchema` and then using [optional](https://hackage.haskell.org/package/base-4.16.1.0/docs/Control-Applicative.html#v:optional) combinator to recover optionality?
Example: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/shelley/exe/cardano-wallet.hs#L204

paolino replied:
I think it's not that  easy 
```haskell
    describe "Tx Metadata JSON option" $ do
        let parse arg = execParserPure defaultPrefs
                (info metadataOption mempty) ["--metadata", arg]
        let md = detailedMetadata
                (TxMetadata (Map.singleton 42 (TxMetaText "hi")))
        let ok ex (Success res) = ex ==  res
            ok _ _ = False
        let err (Failure _) = True
            err _ = False
        mapM_
            (\(desc, arg, tst) -> it desc (parse arg `shouldSatisfy` tst))
            [ ("valid", "{ \"42\": { \"string\": \"hi\" } }", ok (Just md))
            , ("malformed", "testing", err)
            , ("malformed trailing", "{ \"0\": { \"string\": \"\" } } arstneio", err)
            , ("invalid", "{ \"json\": true }", err)
            , ("null 1", "{ \"0\": null }", err)
            , ("null 2", "null", ok Nothing)
            , ("null 3", "{ }", ok (Just (detailedMetadata mempty)))
            ]
```
to pass that `null 2` test we cannot rely on TxMetadataWithSchema

paolino replied:
Can I close this suggestion as a do-nothing, or we can open a ticket to expand the TxMetadataWithSchema to support the `null` value?

piotr-iohk commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2079,14 +2084,15 @@ deleteTransaction ctx (ApiT wid) (ApiTxId (ApiT (tid))) = do
+    -> Maybe Iso8601Time -- ^
+    -> Maybe Iso8601Time -- ^
+    -> Maybe (ApiT SortOrder) -- ^
+    -> Bool -- ^ metadata json schema
```
If I'm not mistaken the new query parameter is not documented in the swagger, so should be added there.

jonathanknowles replied:
I think your PR added some empty Haddock comments 😄 

```suggestion
    => ctx
    -> ApiT WalletId
    -> Maybe MinWithdrawal
    -> Maybe Iso8601Time
    -> Maybe Iso8601Time
    -> Maybe (ApiT SortOrder)
    -> Bool
     -- ^ metadata json schema
```

jonathanknowles replied:
Hi @paolino 

I feel we might be introducing a bit of "Boolean blindness" here.

Specifically, when looking at the type of this parameter, it's not easy to guess that the semantics should be:
- `True`
    _use metadata with no schema_
- `False`
    _use metadata with detailed schema_

Given that you have already introduced the `TxMetadataSchema` type, I think it would be clearer to use this type instead of `Bool` in the type signature of this function.

Call sites then also become more readable, as the reader can explicitly see that this is a parameter of type `TxMetadataSchema`, rather than being just a `True` or `False`.

piotr-iohk commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2105,14 +2119,20 @@ getTransaction
     :: forall ctx s k n. (ctx ~ ApiLayer s k)
     => ctx
     -> ApiT WalletId
+    -> Bool
```
Same here I guess...

paolino replied:
added also, pls check the wording

paolino committed:
add endpoints support for simple metadata

paolino committed:
fix TypesSpec to load the new simple-metadata schema

paolino committed:
fix openapi definitions regarding metadata values

paolino committed:
replace golden test value for transaction cli help

paolino committed:
rm BlockArguments use

sevanspowell committed:
Add a Shelley mnemonic

- Tests were running out of Shelley mnemonics.

paolino committed:
add swagger documentation for get and list transaction parameters
fix documentation for  simple-metadata format

sevanspowell committed:
Fix golden files

paolino committed:
remove Bool switch in transaction CLI to select metadata schema

Anviking committed:
Mark Plutus currency test pending due to flakiness

paolino committed:
Merge pull request #3257 from input-output-hk/jonathanknowles/ADP-1596/simple-metadata

Fixes for ADP-1596

sevanspowell committed:
Add more Shelly mnemonics as a buffer

- Tests are failing due to not enough Shelley mnemonics, add more than necessary
  to stave off future failures.

Anviking committed:
Try adding even more seq test faucets

In response to failures like

>  user error (No more faucet wallet available in MVar!)

seen in CI, even though I don't see them locally.

sevanspowell committed:
Add documentation for no more wallets test error

- Add documentation that describes what to do if the developer encounters a "no
  more wallets" error while testing.

Anviking committed:
Use longer timouet in STAKE_POOLS_SMASH_01

In attempt to alleviate flaky failures.

paolino committed:
refresh golden file ApiTransactionTestnet0.json

piotr-iohk reviewed (approved):
LGTM.

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -1208,7 +1208,10 @@ spec = describe "SHELLEY_STAKE_POOLS" $ do
         \ctx -> runResourceT $ bracketSettings ctx $ do
             updateMetadataSource ctx (_smashUrl ctx)
-            eventually "metadata is fetched" $ do
+            -- This can be slow; let's retry more seldom but with a longer
```
```suggestion
            -- This can be slow; let's retry less frequently but with a longer
```

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -972,6 +977,28 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
+
+        basePayload <- mkTxPayload ctx wb amt fixturePassphrase
+
+        let txMeta = [json|{ "1": #{T.replicate 65 "a"}    }|]
```
```suggestion
        let txMeta = [json|{ "1": #{T.replicate 65 "a"} }|]
```

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -1020,7 +1050,51 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
+
+        payload <- mkTxPayload ctx wb amt fixturePassphrase
+
+        let txMeta = [json|{ "1": "hello"    }|]
```
```suggestion
        let txMeta = [json|{ "1": "hello" }|]
```

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -1035,6 +1109,28 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
+
+        basePayload <- mkTxPayload ctx wb amt fixturePassphrase
+
+        let txMeta = [json|{ "1":    #{T.replicate 65 "a" } }|]
```
```suggestion
        let txMeta = [json|{ "1": #{T.replicate 65 "a" } }|]
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs:
```diff
@@ -0,0 +1,71 @@
+        ( fmap noSchemaMetadata
+                . either (fail . displayError) pure
+                . metadataFromJson TxMetadataJsonNoSchema
+        )
```
(4-space indents.)

```suggestion
        (fmap detailedMetadata
            . either (fail . displayError) pure
            . metadataFromJson TxMetadataJsonDetailedSchema
        )
        (fmap noSchemaMetadata
            . either (fail . displayError) pure
            . metadataFromJson TxMetadataJsonNoSchema
        )
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -2157,6 +2159,11 @@ instance Arbitrary (PostTransactionOldData n) where
 
+instance Arbitrary TxMetadataWithSchema where
+  arbitrary = TxMetadataWithSchema
+    <$> elements [TxMetadataNoSchema, TxMetadataDetailedSchema]
```
It might be worth considering the use of `arbitraryBoundedEnum` here, if you can derive `Bounded` and `Enum` for this type. Then we'd automatically get coverage of new constructors that are added. (If we add them.)

```suggestion
    <$> arbitraryBoundedEnum
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Link.hs:
```diff
@@ -676,15 +684,19 @@ getTransaction
-    mkURL mk = mk wid tid
+    mkByronURL mk = mk wid tid
+
+    mkShelleyURL :: (ApiT WalletId -> Bool  -> ApiTxId -> Text) -> Text
```
```suggestion
    mkShelleyURL :: (ApiT WalletId -> Bool -> ApiTxId -> Text) -> Text
```

jonathanknowles committed:
Formatting fixes.

Also remove blank haddock comments.

jonathanknowles committed:
Tweak language in integration test comments.

jonathanknowles committed:
Use `TxMetadataSchema` type in `{getTransactions,listTransactions}`.

We use this type instead of `Bool`, which can obscure the intended
semantics.

jonathanknowles committed:
Use `TxMetadataSchema` type in `Api.Link.listTransactions`.

Explicitly indicate the type of metadata schema required.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Link.hs:
```diff
@@ -676,15 +684,19 @@ getTransaction
         )
     => w
     -> t
+    -> Bool
```
Hi @paolino 

Similarly to my previous comment, I feel this is introducing a bit of "boolean blindness".

If someone reading the integration tests for the first time sees `True` or `False` at call sites, they may wonder what the expected semantics are. To find out, they'd (presumably) navigate to the `Api.Link.getTransaction` function, but then be none the wiser, since the `Bool` value is simply passed to the inner function. They'd have to go all the way through the API, and then into the definition of `Api.Server.getTransaction` to figure out that:

- `True`
    maps to `TxMetadataNoSchema`
- `False`
    maps to `TxMetadataDetailedSchema`

This comprehension barrier might be compounded if we have multiple Boolean flags.

If it is necessary for callers of `Api.Link.getTransaction` to specify this flag, then I think we'd benefit from using the `TxMetadataSchema` type here. Readers of call sites would then be able to navigate to the definition of `TxMetadataSchema` and figure out the intended semantics by searching for uses of `TxMetadata` through the code base.

jonathanknowles committed:
Remove metadata schema parameter from `Api.Link.getTransaction`.

This function is always called with a value of `False` for the "simple
metadata" parameter. Therefore, we can remove this parameter from the
function and use a constant, which removes the burden of including the
parameter at all call sites.

To avoid Boolean blindness, we specify the `TxMetadataDetailedSchema`
constructor explicitly (which is equivalent to simple metadata = `False`).

jonathanknowles reviewed (changes requested):
Hi @paolino 

Thanks for making this PR again!

I've given it a more thorough review, and it looks mostly good to me.

Though I can see two issues:

1. I have reservations about the use of `Bool` in certain function type signatures. I think this may contribute to "Boolean blindness", particularly at call sites, where I think it might be difficult for readers to determine what the intended semantics are (see review comments). Given that we have a rich `TxMetadataSchema` type available, I think we can use that instead in most cases. (This also applies to integration tests.)

2. We currently don't have an explicit JSON roundtrip test for `TxMetadataWithSchema`.

I've been preparing a small PR with proposed fixes. I'll let you know when it's ready to review. (**UPDATE:** See https://github.com/input-output-hk/cardano-wallet/pull/3253#issuecomment-1121931035)

Thanks again!

Jonathan

jonathanknowles committed:
Use consistent parameter order in CLI and API for `getTransaction`.

We move the `simple-metadata` parameter to the end.

The other parameters are more fundamental, and appear first.

jonathanknowles committed:
Use `TxMetadataSchema` type in `Api.Client`.

jonathanknowles committed:
Use `TxMetadataSchema` type in `listTransactionsViaCLI`.

jonathanknowles committed:
Use `TxMetadataSchema` in `getTransactionViaCLI`.

jonathanknowles committed:
Add JSON roundtrip test for `TxMetadataWithSchema`.

jonathanknowles committed:
Merge pull request #3270 from input-output-hk/jonathanknowles/ADP-1596/simple-metadata

Use `TxMetadataSchema` in preference to `Bool` for function parameters.

paolino committed:
remove duplicate MetadataSchema datatype

jonathanknowles reviewed (approved):
LGTM!

HeinrichApfelmus commented on lib/core-integration/src/Test/Integration/Scenario/CLI/Shelley/Transactions.hs:
```diff
@@ -317,34 +319,105 @@ spec = describe "SHELLEY_CLI_TRANSACTIONS" $ do
+        \Transaction with metadata via CLI with simple metadata" $
+        \ctx -> runResourceT $ do
+
+        (wSrc, wDest) <- (,) <$> fixtureWallet ctx <*> emptyWallet ctx
```
(Tip for the future: Avoids operator noise:)

```suggestion
        wSrc <- fixtureWallet ctx
        wDest <- emptyWallet ctx
```

HeinrichApfelmus commented on lib/core-integration/src/Test/Integration/Scenario/CLI/Shelley/Transactions.hs:
```diff
@@ -418,7 +492,8 @@ spec = describe "SHELLEY_CLI_TRANSACTIONS" $ do
                     (Exit code, Stdout out, Stderr err) <-
-                        listTransactionsViaCLI ctx $ join
+                        listTransactionsViaCLI ctx TxMetadataDetailedSchema $
+                        join
```
Indentation of `join` is misleading — it's an argument to `listTransactionsViaCLI` and merits further indentation. The position on the previous line was better in that regard.

HeinrichApfelmus commented on lib/core/src/Cardano/Wallet/Api/Client.hs:
```diff
@@ -333,7 +337,9 @@ transactionClient =
-            , getTransaction = _getTransaction
+            , getTransaction =
+                \wid txid metadataSchema ->
+                _getTransaction wid txid (toSimpleMetadataFlag metadataSchema)
```
I wasn't quite sure about the `do` notation in other places, but I also have the feeling here that the body of the lambda should be indented relative to its head.

paolino replied:
yep, this is why I really would like to have an autoformatter. I'm not indenting anything by hand since quite some time

HeinrichApfelmus commented on lib/core/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs:
```diff
@@ -0,0 +1,93 @@
+import Prelude
+
+-- | a tag to select the json codec
+data TxMetadataSchema = TxMetadataNoSchema | TxMetadataDetailedSchema
```
👍

HeinrichApfelmus commented on lib/core/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs:
```diff
@@ -0,0 +1,93 @@
+-- | a wrapper to drive the json codec of metadata
+data TxMetadataWithSchema = TxMetadataWithSchema
+    { -- | how to codec the metadata into json
+        txMetadataWithSchema_schema :: TxMetadataSchema
```
Excellent idea 👍 — same data, but a flag indicating how it should be encoded.

HeinrichApfelmus commented on lib/core/src/Cardano/Wallet/Api/Types/SchemaMetadata.hs:
```diff
@@ -0,0 +1,93 @@
+data TxMetadataSchema = TxMetadataNoSchema | TxMetadataDetailedSchema
+    deriving (Show, Eq, Generic, NFData)
+
+-- | a wrapper to drive the json codec of metadata
```
(Tip for future: Treating haddock comments as full sentences looks better in the rendered documentation. Hence, starting with an upper case is preferred.)

paolino replied:
fixed

paolino committed:
Update lib/core-integration/src/Test/Integration/Scenario/CLI/Shelley/Transactions.hs

Co-authored-by: Heinrich Apfelmus <heinrich.apfelmus@iohk.io>

paolino committed:
apply stylistic suggestions
